File: packages/shared/lib/breakthrough/breakthrough-shift-detector.ts
import type { SessionMovementKind } from "@/types/session-movement-summary";

/** Minimum confidence swing to surface a breakthrough card in feeds. */
export const BREAKTHROUGH_CONFIDENCE_DELTA_THRESHOLD = 10;

const BREAKTHROUGH_MOVEMENT_KINDS = new Set<SessionMovementKind>([
  "belief_changed",
  "belief_weakened",
  "belief_strengthened",
  "contradiction_appeared",
]);

export interface BreakthroughShiftInput {
  confidenceDelta?: number;
  movementKind?: SessionMovementKind | null;
  statusChanged?: boolean;
}

export function isBreakthroughShift(input: BreakthroughShiftInput): boolean {
  if (input.statusChanged) return true;
  if (
    input.movementKind != null &&
    BREAKTHROUGH_MOVEMENT_KINDS.has(input.movementKind)
  ) {
    return true;
  }
  const delta = Math.abs(input.confidenceDelta ?? 0);
  return delta >= BREAKTHROUGH_CONFIDENCE_DELTA_THRESHOLD;
}

function parseConfidenceDeltaFromDetail(detailLine?: string): number | undefined {
  if (!detailLine) return undefined;
  const match = detailLine.match(/(\d+)%\s*→\s*(\d+)%/);
  if (!match) return undefined;
  return Math.abs(Number(match[2]) - Number(match[1]));
}

/** Maps a post-save session movement summary to breakthrough feed eligibility. */
export function isBreakthroughFromSessionMovement(input: {
  kind: SessionMovementKind;
  detailLine?: string;
}): boolean {
  return isBreakthroughShift({
    movementKind: input.kind,
    statusChanged: input.kind === "belief_changed",
    confidenceDelta: parseConfidenceDeltaFromDetail(input.detailLine),
  });
}

export const BREAKTHROUGH_FEED_COPY = {
  eyebrow: "Breakthrough",
  title: "Something meaningful shifted in your archive",
  body: "This is a high-signal change — worth revisiting when you have a quiet moment.",
} as const;

File: apps/mobile/retired_sprawl/lib_features/breakthrough_tracking/breakthrough_shift_detector.dart
import 'package:archiveme_mobile/features/session_movement/session_movement_models.dart';

abstract final class BreakthroughShiftDetector {
  BreakthroughShiftDetector._();

  static const confidenceDeltaThreshold = 10;

  static bool isBreakthroughMovement(SessionMovementKind kind) {
    return switch (kind) {
      SessionMovementKind.beliefChanged ||
      SessionMovementKind.beliefWeakened ||
      SessionMovementKind.beliefStrengthened ||
      SessionMovementKind.contradictionAppeared =>
        true,
      _ => false,
    };
  }

  static bool isBreakthroughSummary(SessionMovementSummaryView summary) {
    if (summary.kind == SessionMovementKind.beliefChanged) return true;
    if (isBreakthroughMovement(summary.kind)) return true;
    final delta = _parseConfidenceDelta(summary.detailLine);
    return delta != null && delta >= confidenceDeltaThreshold;
  }

  static int? _parseConfidenceDelta(String? detailLine) {
    if (detailLine == null) return null;
    final match = RegExp(r'(\d+)%\s*→\s*(\d+)%').firstMatch(detailLine);
    if (match == null) return null;
    final from = int.tryParse(match.group(1) ?? '');
    final to = int.tryParse(match.group(2) ?? '');
    if (from == null || to == null) return null;
    return (to - from).abs();
  }
}

abstract final class BreakthroughFeedCopy {
  BreakthroughFeedCopy._();

  static const eyebrow = 'Breakthrough';
  static const title = 'Something meaningful shifted in your archive';
  static const body =
      'This is a high-signal change — worth revisiting when you have a quiet moment.';
}